presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,285 @@
+    public void start()
+    {
+        if (backupAvailable && coordinator && !started.getAndSet(true)) {
+            startJob();
+        }
+    }
+
+    @PreDestroy
+    public void shutdown()
+    {
+        executor.shutdownNow();
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getBucketsBalanced()
+    {
+        return bucketsBalanced;
+    }
+
+    @Managed
+    @Nested
+    public CounterStat getJobErrors()
+    {
+        return jobErrors;
+    }
+
+    private void startJob()
+    {
+        executor.scheduleWithFixedDelay(() -> {
+            try {
+                process();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error balancing buckets");
+                jobErrors.update(1);
+            }
+        }, 0, interval.toMillis(), MILLISECONDS);
+    }
+
+    @VisibleForTesting
+    int process()
+    {
+        Set<String> activeNodes = nodeSupplier.getWorkerNodes().stream()
+                .map(Node::getNodeIdentifier)
+                .collect(Collectors.toSet());
+
+        Map<String, Double> allocationBytes = new HashMap<>(activeNodes.stream().collect(Collectors.toMap(node -> node, node -> 0.0)));
+        Multimap<String, BucketAssignment> sourceToAllocationChanges = HashMultimap.create();
+
+        for (Distribution distribution : shardManager.getDistributions()) {
+            Multimap<String, BucketAssignment> existingAllocations = HashMultimap.create();
+            Multiset<String> allocationCounts = HashMultiset.create();
+
+            List<BucketNode> buckets = shardManager.getBucketNodes(distribution.getId());
+            double bucketSize = (1.0 * shardManager.getDistributionSizeBytes(distribution.getId())) / buckets.size();
+
+            for (BucketNode bucketNode : buckets) {
+                String node = bucketNode.getNodeIdentifier();
+                existingAllocations.put(node, assignment(distribution.getId(), bucketNode.getBucketNumber(), node));
+                allocationCounts.add(node);
+
+                allocationBytes.compute(node, (k, v) -> v == null ? bucketSize : v + bucketSize);
+            }
+
+            int currentMin = allocationBytes.keySet().stream()
+                    .map(allocationCounts::count)
+                    .min(Integer::compare)
+                    .get();
+            int currentMax = allocationBytes.keySet().stream()
+                    .map(allocationCounts::count)
+                    .max(Integer::compare)
+                    .get();
+
+            int targetMin = (int) Math.floor((existingAllocations.size() * 1.0) / activeNodes.size());
+            int targetMax = (int) Math.ceil((existingAllocations.size() * 1.0) / activeNodes.size());
+
+            log.info("Distribution %s: Current bucket skew: min %s, max %s. Target bucket skew: min %s, max %s", distribution.getId(), currentMin, currentMax, targetMin, targetMax);
+
+            for (String source : ImmutableSet.copyOf(allocationCounts)) {
```

Comment:
Currently when a node is not available, its buckets are [reassigned randomly](https://github.com/prestodb/presto/blob/master/presto-raptor/src/main/java/com/facebook/presto/raptor/metadata/DatabaseShardManager.java#L673). The node selection process used by the balancer should also be used there.